src/encoder.rs

```diff
@@ -0,0 +1,236 @@
+fn write_sample(name: &str,
+                mc: &proto::Metric,
+                additional_label_name: &str,
+                additional_label_value: &str,
+                value: f64,
+                writer: &mut Write)
+                -> Result<usize> {
+    let mut written = 0;
+    written += try!(writer.write(name.as_bytes()).or_else(|e| Err(Error::Io(e))));
+
+    written += try!(label_pairs_to_text(mc.get_label(),
+                                        additional_label_name,
+                                        additional_label_value,
+                                        writer));
+
+    written += try!(writer.write((format!(" {}", value)).as_bytes())
+        .or_else(|e| Err(Error::Io(e))));
+    let time_stamp = mc.get_timestamp_ms();
+    if time_stamp != 0 {
+        written += try!(writer.write(format!(" {}", time_stamp).as_bytes())
+            .or_else(|e| Err(Error::Io(e))));
+    }
+
+    written += try!(writer.write(b"\n").or_else(|e| Err(Error::Io(e))));;
+    Ok(written)
+}
+
+/// `label_pairs_to_text` converts a slice of `LabelPair` proto messages plus the
+/// explicitly given additional label pair into text formatted as required by the
+/// text format and writes it to 'out'. An empty slice in combination with an
+/// empty string `additional_label_name` results in nothing being
+/// written. Otherwise, the label pairs are written, escaped as required by the
+/// text format, and enclosed in '{...}'. The function returns the number of
+/// bytes written and any error encountered.
+fn label_pairs_to_text(pairs: &[proto::LabelPair],
+                       additional_label_name: &str,
+                       additional_label_value: &str,
+                       writer: &mut Write)
+                       -> Result<usize> {
+    if pairs.len() == 0 && additional_label_name == "" {
+        return Ok(0);
+    }
+
+    let mut written = 0;
+
+    let mut separator = "{";
+    for lp in pairs {
+        written += try!(writer.write((format!("{}{}=\"{}\"",
+                            separator,
+                            lp.get_name(),
+                            escape_string(lp.get_value(), true)))
+                .as_bytes())
+            .or_else(|e| Err(Error::Io(e))));
+
+        separator = ",";
+    }
+
+    if additional_label_name != "" {
+        written += try!(writer.write((format!("{}{}=\"{}\"",
+                            separator,
+                            additional_label_name,
+                            escape_string(additional_label_value, true)))
+                .as_bytes())
+            .or_else(|e| Err(Error::Io(e))));
+    }
+
+    written += try!(writer.write(b"}").or_else(|e| Err(Error::Io(e))));
+    Ok(written)
+}
+
+/// `escape_string` replaces '\' by '\\', new line character by '\n', and - if
+/// `include_double_quote` is true - '"' by '\"'.
+pub fn escape_string(v: &str, include_double_quote: bool) -> String {
```

Comment:
If there is no standard for escaping, I prefer just using [char#escape_default](https://doc.rust-lang.org/stable/std/primitive.char.html#method.escape_default)